README.md
Financial Regulatory Transparency Index
========

**Version:** 0.3.3

**Maintainer:**
[Christopher Gandrud](http://christophergandrud.blogspot.de/p/biocontact.html)

Funding generously provided by the
[Deutsche Forschungsgemeinschaft](http://www.dfg.de/en/).

**Work In Progress**

---

## Motivation

Why do countries release data on their financial systems to international
organizations, such as the IMF and World Bank? What are the consequences of
releasing this data for the stability of their financial systems?

To address these questions we are developing a **Financial Regulatory
Transparency Index**. The new Index we make it possible to compare the
willingness of governments to credibly reveal the structure of their financial
systems through international institutions, allowing them to be scrutinized by
market participants and citizens.

We presented an early draft of a research using the FRT at the 2015 Political
Economy of International Organizations conference. The working paper can be
found
[here](http://wp.peio.me/wp-content/uploads/PEIO8/Copelovitch,%20Gandrud,%20Hallerberg%203.2.2015.pdf).

## The FRT Index

The current draft version of the Index is located in the *IndexData* directory
in a CSV formatted file called:
[FRTIndex.csv](https://raw.githubusercontent.com/FGCH/FRTIndex/master/IndexData/FRTIndex.csv).

It covers the **68 countries** classified by the World Bank as
'**High Income**', [EMBI](http://en.wikipedia.org/wiki/JPMorgan_EMBI) countries,
and China for the years **1990 through 2011**.

#### FRT Index Scores for selected years

![FRT Overview](FRT_overview.png)

The file `FRTIndex.csv` contains the following variables:

| Variable Name | Short Description                                          |
| ------------- | ---------------------------------------------------------- |
| country       | country name                                               |
| iso2c         | [ISO 2 letter country code](http://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) |
| year          | year of the FRT score                                      |
| lower_95      | lower bound of the 95% highest probability density interval|
| lower_90      | lower bound of the 90% highest probability density interval|
| median        | median of the FRT index posterior distribution             |
| upper_90      | upper bound of the 90% highest probability density interval|
| upper_95      | upper bound of the 95% highest probability density interval|

## Download into R

To download the **working version** of the Index directly into
[R](http://www.r-project.org/) as a data frame use:

```{S}
URL <- 'https://raw.githubusercontent.com/FGCH/FRTIndex/master/IndexData/FRTIndex.csv'

frt_index <- rio::import(URL)
```

## Estimation Model

The FRT Index is created using a
[Bayesian Item Response Theory](http://en.wikipedia.org/wiki/Item_response_theory)
model of high income countries's reporting of [financial industry indicators](https://github.com/FGCH/FRTIndex/blob/master/source/IndicatorDescript/IncludedIndicators.csv)
to the World Bank's
[Global Financial Development Database](http://data.worldbank.org/data-catalog/global-financial-development).

A full write up of our model is in the works. The most recent (incomplete) draft
is available for download as a
[PDF](https://github.com/FGCH/FRTIndex/blob/master/paper/FRTIndexPaper.pdf?raw=true).

Our estimation model is based on:

> Hollyer, James R., B. Peter Rosendorff, and James Raymond Vreeland. 2014.
"Replication data for: Measuring Transparency".
[http://dx.doi.org/10.7910/DVN/24274](http://dx.doi.org/10.7910/DVN/24274)

and

> [Stan Modeling Language: User's Guide and Reference Manual](http://mc-stan.org/manual.html)
for Stan Version 2.5.0. p. 49-59.

We have built on this model using the
[No-U-Turn Sampler](http://arxiv.org/abs/1111.4246) implemented in
[Stan](http://mc-stan.org/).

---

<a href="http://www.dfg.de/en/"><img src="http://fgch.github.io/amc-site/img/dfg.png"/></a> <a href="http://nadrosia.tumblr.com/post/53520500877/made-in-berlin-badge-update"><img alt="image" src="http://media.tumblr.com/023c285c14ef01953d3b67ffe789004d/tumblr_inline_mor1uu2OOZ1qz4rgp.png" height = "50"></a> <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="http://i.creativecommons.org/l/by-sa/4.0/88x31.png" height = "40" /></a>


changelog.md
# Changelog

## Version 0.3.2

- Run with 80,000 iterations. Thin = 20 due to memory limitations.

## Version 0.3.1

- Run with 40,000 iterations. Thin = 10 due to memory limitations.

## Version 0.3

- Includes EMBI countries + China.

- Run with 20,000 iterations.

## Version 0.2.5

- Dropped Liquid liabilities to GDP (%) (GFDD.DI.05) as it is very similar to
Liquid liabilities in millions USD (2000 constant).

- Modified Stan code so that gamma is drawn from a Log-Normal distribution,
rather than exponentiating gamma in the logistic model. Gamma is also explicitly
constrained to be lower bounded by 0.

## Version 0.2.4.1

Same as v0.2.4, but highest probability density intervals reported rather than
central intervals.

## Version 0.2.4

- Run with 10,000 iterations.

## Version 0.2.3

- Run with 6,000 iterations.

## Version 0.2.2

- Use cauchy(0,0.05) priors for transparency scale parameter and mean
transparency. A smaller scale parameter (previously it was 0.25) is  a more
informative prior used to limit the range of the transparency parameter.

- Run with 3,000 iterations.

## Version 0.2.1

- Fill in data missing from the World Bank's version of the GFDD
with data from FRED's version of the GFDD.

- No longer include jurisdictions that never report items.
So, now there are 50 countries in the sample.

## Version 0.2

Updated beta model FRT Index using the No-U-Turn Sampler in Stan.

- 14 items

- 1990-2011.

- Model run with 4 chains with 2,000 iterations each (1000 or which are
burn-in).

## Version 0.1

Initial beta model FRT Index with 60 countries, 21 items, and 12 yeas
(1998-2011). The model was run with 2 chains, 6,000 iterations with a 5,000
iteration burn-in period using JAGS.


source/GFDD_Base.R
##############
# Play with Global Financial Development Data
# Christopher Gandrud
# 18 March 2014
#############

## Inspired by:
# Hollyer, James R. ; Rosendorff, B. Peter; Vreeland, James Raymond, 2014, 
# "Replication data for: Measuring Transparency", 
# http://dx.doi.org/10.7910/DVN/24274 UNF:5:Mj2tZ0rivXJPhuFqdQS0JA== IQSS Dataverse Network 
# [Distributor] V1 [Version]

setwd('~/FRTOutFiles/')

# Load packages
library(WDI)
library(DataCombine)
# library(arm)
library(rjags)
library(R2jags)
library(xtable)

#### Create Indicator Data Set ####
# Download GFDD data
Indicators <- c('GFDD.DI.01', 'GFDD.DI.02', 'GFDD.DI.03', 'GFDD.DI.04', 'GFDD.DI.05', 'GFDD.DI.06',
                'GFDD.DI.07', 'GFDD.DI.08', 'GFDD.DI.11', 'GFDD.DI.12', 'GFDD.DI.13', 'GFDD.DI.14', 
                'GFDD.EI.02', 'GFDD.EI.08', 'GFDD.OI.02', 'GFDD.OI.07', 'GFDD.OI.13', 'GFDD.SI.02',
                'GFDD.SI.03', 'GFDD.SI.04', 'GFDD.SI.05', 'GFDD.SI.07') 

# Download indicators
# Unable to download 'GFDD.DM.011', 'GFDD.OI.14'
Base <- WDI(indicator = Indicators, start = 1998, end = 2011, extra = TRUE)

# Keep countries with 'High income' (OECD and non-OECD classification)
BaseSub <- grepl.sub(data = Base, Var = 'income', patterns = 'High income')
Droppers <- c("iso3c", "region",  "capital", "longitude", "latitude", "income", "lending")
BaseSub <- BaseSub[, !(names(BaseSub) %in% Droppers)]

# Create missingness indicators
KeeperLength <- length(Indicators)
IndSub <- Indicators[1:KeeperLength]
VarVec <- vector()

for (i in IndSub){
  BaseSub[, paste0('Rep_', i)] <- 1
  BaseSub[, paste0('Rep_', i)][is.na(BaseSub[, i])] <- 0
  
  temp <- paste0('Rep_', i)
  VarVec <- append(VarVec, temp)
}

# Create country numbers
BaseSub$countrynum <- as.numeric(as.factor(BaseSub$iso2c)) 
BaseSub$yearnum <- as.numeric(as.factor(BaseSub$year)) 

# Create keys
CountryKey <- BaseSub[, c('countrynum', 'iso2c', 'country')]
CountryKey <- CountryKey[!duplicated(CountryKey$countrynum), ]

YearKey <- BaseSub[, c('yearnum', 'year')]
YearKey <- YearKey[!duplicated(YearKey$yearnum), ]

if (getwd() == "/git_repositories/FRTIndex/source"){
  IndicatorKey <- read.csv('IndicatorDescript/IndicatorDescription.csv', encoding = 'latin1', stringsAsFactors = FALSE)
  IndicatorKey <- subset(IndicatorKey, SeriesCode %in% IndSub)
  write.csv(IndicatorKey, file = 'IndicatorDescript/IncludedIndicators.csv', row.names = FALSE)
  
  # Clean up for latex version 
  IndicatorKey$Source <- gsub(pattern = 'International Financial Statistics \\(IFS\\), International Monetary Fund \\(IMF\\)', 
                             replacement = 'IFS/IMF', x = IndicatorKey$Source)
  IndicatorKey$Source <- gsub(pattern = 'Financial Soundness Indicators Database \\(fsi\\.imf\\.org\\), International Monetary Fund \\(IMF\\)', 
                              replacement = 'IFSI/IMF', x = IndicatorKey$Source) 
  IndicatorKey$Source <- gsub(pattern = 'World Development Indicators \\(WDI\\), World Bank', 
                              replacement = 'World Bank', x = IndicatorKey$Source)  
  IndicatorKey$Source <- gsub(pattern = 'World Bank - Non banking financial database', 
                              replacement = 'World Bank', x = IndicatorKey$Source)  
  IndicatorKey$Source <- gsub(pattern = 'Nonbanking financial database, World Bank', 
                              replacement = 'World Bank', x = IndicatorKey$Source)  
  IndicatorKey$Periodicity <- gsub('Annual: ', '', IndicatorKey$Periodicity)
  
  print(xtable(IndicatorKey), size = 'scriptsize', include.rownames = FALSE, floating = FALSE, 
        file = '/git_repositories/FRTIndex/paper/tables/IndicatorDescript.tex')
}

# Keep only complete variables
BaseJagsReady <- BaseSub[, c('countrynum', 'yearnum', VarVec)]

# Create vector of parameters to estimate
VarCount <- 1:length(VarVec)
Betas <- paste0('beta', VarCount)
ParamsEst <- c("transparency", "tau", Betas)
Num <- nrow(BaseJagsReady)
NCountry <- max(BaseJagsReady$countrynum)
Nyear <- max(BaseJagsReady$yearnum)

#### Write JAGS model ####
Xs <- as.character()
Ps <- as.character()
Qs <- as.character()
Vs <- as.character()
Bs <- as.character()

for (n in VarCount){
  temp <- paste0('x', n, '[n] <- beta', n, '[1] + transparency[countrynum[n], yearnum[n]]*beta', n,'[2]')
  Xs <- paste(Xs, temp, sep = '\n') 
  
  temp <- paste0('p', n, '[n] <- 1/(1 + exp(-x', n, '[n]))')
  Ps <- paste(Ps, temp, sep = '\n')  
  
  temp <- paste0('q', n, '[n] <- max(min(p', n, '[n],1),0)')  
  Qs <- paste(Qs, temp, sep = '\n')
  
  temp <- paste0(VarVec[n], '[n]~dbern(q', n, '[n])')
  Vs <- paste(Vs, temp, sep = '\n')
  
  temp <- paste0('beta', n, '[1:2] ~ dmnorm(mu[1:2],alpha[1:2,1:2])')
  Bs <- paste(Bs, temp, sep = '\n')
  
}

cat(paste0('
model{
    for (n in 1:', Num, '){\n',
      Xs, '\n',
      Ps, '\n',
      Qs, '\n',
      Vs,
    '\n }',  
  '\n# Model priors\n',    
    '
mu[1] <- 0
mu[2] <- 0

alpha[1,1] <- 0.01
alpha[1,2] <- 0
alpha[2,1] <- 0
alpha[2,2] <- 0.01\n',

Bs,

'\n\n# Transparency priors\n',

paste0('
  for (n in 1:', NCountry, '){
    transinit[n] ~ dnorm(0, 0.01)
  }
  mean.trans <- mean(transinit[1:', NCountry,'])
  sd.trans <- sd(transinit[1:', NCountry, '])

  for (n in 1:', NCountry, '){
    transcentered[n] <- (transinit[n] - mean.trans)*pow(sd.trans, 1)
  }

  for (n in 1:', NCountry, '){
    transparency[n,1] <-transcentered[n]
    tau[n] ~ dgamma(20, 4)
  
    for (j in 2:', Nyear, '){
      transparency[n,j] ~ dnorm(transparency[n,(j-1)], tau[n])
    }
  }'), 
'\n}'), file = 'BasicModel_V1.bug')

# Copy file into git repo for version control
file.copy(from = 'BasicModel_V1.bug',
          to = '/home/cjg/FRTIndex/source/BasicModel_V1.bug')

#### Run JAGS Model #### 
# Create list of data objects used by the model
# DataList <- append(list('countrynum', 'year'), as.list(VarVec))

DataList <- list('countrynum' = BaseJagsReady$countrynum, 'yearnum' = BaseJagsReady$yearnum, 
                 'Rep_GFDD.DI.01' = BaseJagsReady$Rep_GFDD.DI.01, 
                 'Rep_GFDD.DI.02' = BaseJagsReady$Rep_GFDD.DI.02, 'Rep_GFDD.DI.03' = BaseJagsReady$Rep_GFDD.DI.03,
                 'Rep_GFDD.DI.04' = BaseJagsReady$Rep_GFDD.DI.04, 'Rep_GFDD.DI.05' = BaseJagsReady$Rep_GFDD.DI.05, 
                 'Rep_GFDD.DI.06' = BaseJagsReady$Rep_GFDD.DI.06, 'Rep_GFDD.DI.07' = BaseJagsReady$Rep_GFDD.DI.07, 
                 'Rep_GFDD.DI.08' = BaseJagsReady$Rep_GFDD.DI.08, 'Rep_GFDD.DI.11' = BaseJagsReady$Rep_GFDD.DI.11,
                 'Rep_GFDD.DI.12' = BaseJagsReady$Rep_GFDD.DI.12, 'Rep_GFDD.DI.13' = BaseJagsReady$Rep_GFDD.DI.13,
                 'Rep_GFDD.DI.14' = BaseJagsReady$Rep_GFDD.DI.14, 'Rep_GFDD.EI.02' = BaseJagsReady$Rep_GFDD.EI.02,
                 'Rep_GFDD.EI.08' = BaseJagsReady$Rep_GFDD.EI.08, 'Rep_GFDD.OI.02' = BaseJagsReady$Rep_GFDD.OI.02,
                 'Rep_GFDD.OI.07' = BaseJagsReady$Rep_GFDD.OI.07, 'Rep_GFDD.OI.13' = BaseJagsReady$Rep_GFDD.OI.13, 
                 'Rep_GFDD.SI.02' = BaseJagsReady$Rep_GFDD.SI.02, 'Rep_GFDD.SI.03' = BaseJagsReady$Rep_GFDD.SI.03,
                 'Rep_GFDD.SI.04' = BaseJagsReady$Rep_GFDD.SI.04, 'Rep_GFDD.SI.05' = BaseJagsReady$Rep_GFDD.SI.05,
                 'Rep_GFDD.SI.07' = BaseJagsReady$Rep_GFDD.SI.07)

# Betas <- paste0('beta', VarCount)
parameters <- c("transparency", "tau", Betas)

# Send to JAGS
# Est1 <- jags(data = DataList, inits = NULL, parameters, model.file = "BasicModel1.bug",
#             n.chains = 2, n.iter = 1000, n.burnin = 50)

Est1 <- jags.model('BasicModel_V1.bug', data = DataList, 
                   n.chains = 2, n.adapt = 1000)

#save(Est1, file = 'ModelPlay.rda')
load('ModelPlay.rda')

update(Est1, 5000)

#save(Est1, file = 'ModelPlay2.rda')

#### Examine Simulations ####

Test <- coda.samples(Est1, parameters, 1000)

#save(Test, file = 'ModelPlay3.rda')



